File: Cpp/Medium/347. 前K个高频元素.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> map1;
        vector<vector<int>> bucket(nums.size() + 1);
        vector<int> res;
        for(auto a : nums) {
            map1[a]++;
        }
        for(auto it : map1) {
            bucket[it.second].push_back(it.first);
        }
        for(int i = nums.size(); i >= 0; i--) {
            for(int j = 0; j < bucket[i].size(); j++) {
                res.push_back(bucket[i][j]);
                if(res.size() == k) {
                    return res;
                }
            }
        }
        return res;
    }
};

File: Java/medium/347. 前K个高频元素.java
class Solution {
    public List<Integer> topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        List<Integer>[] bucket = new ArrayList[nums.length + 1];
        for(int key : map.keySet()) {
            int frequency = map.get(key);
            if(bucket[frequency] == null) {
                bucket[frequency] = new ArrayList<>();
            }
            bucket[frequency].add(key);
        }
        List<Integer> top = new ArrayList<>();
        for(int i = bucket.length - 1; i >= 0 && top.size() < k; i--) {
            if(bucket[i] == null) {
                continue;
            }
            if(bucket[i].size() <= (k - top.size())) {
                top.addAll(bucket[i]);
            } else {
                top.addAll(bucket[i].subList(0, k - top.size()));
            }
        }
        return top;
    }
}
